ql/instrument.hpp

```diff
@@ -200,6 +201,22 @@ namespace QuantLib {
         return additionalResults_;
     }
 
+    //! Utility function to optimize the observability graph of an instrument
+    /*! This function unregisters the given instrument (e.g. a Swap) from the given cashflows and
+        instead registers with the observables of the cashflows. This is safe to do if
+        - the coupon pricers of the cashflows are set before the function is called and never
+          updated afterwards
+        - the cashflows are not themselves originating notifications, i.e. they only pass through
+          notifications from their observables (which is usually the case)
+        - the set of cashflows does not dynamically change (usually satisfied as well)
+        If unregisterCoupons is set to true, all given cashflows are in addition unregistered from
+        all their observables. This can be done
+        - if the coupons are not asked for results directly
+        - if deepUpdate() is called on the instrument before retrieving a result; to determine
+          whether the result might have changed, isCalculated() can be called on the instrument.
+    */
+    void passThroughNotifications(Instrument& instrument,
+                                  const std::vector<Leg>& legs,
```

Comment:
For the basic function I'd pass just one leg here, so as not to force instruments such as bonds to provide a vector of legs.  Then we can also overload it for multiple legs, or (maybe even better) add overloads such as e.g. `passThroughNotifications(Swap& swap, bool unregisterCoupons = false);` that would loop over the swap legs and call the basic function.

Reply:
I separated the whole stuff out to a new source file, that's probably better